router/h2/src/main/scala/io/buoyant/router/h2/ClassifiedRetryFilter.scala

```diff
@@ -208,27 +212,57 @@ class ClassifiedRetryFilter(
   }
 
   /** Read and release all frames from the Stream */
-  private[this] def consumeAll(stream: Stream): Future[Unit] = {
-    consumeAllButLast(stream).map {
+  private[this] def consumeAll(stream: Stream, deadline: Option[Time] = None): Future[Unit] = {
+    consumeAllButLast(stream, deadline).map {
       case Some(frame) =>
         frame.release(); ()
       case None => ()
     }
   }
 
+  private[this] val classificationTimeoutException = new TimeoutException("Classification timeout")
+
   /** Read and release all but the last frame from a Stream.  Then return the last frame */
-  private[this] def consumeAllButLast(stream: Stream): Future[Option[Frame]] =
+  private[this] def consumeAllButLast(stream: Stream, deadline: Option[Time]): Future[Option[Frame]] =
     if (stream.isEmpty) {
       Future.None
     } else {
-      stream.read().flatMap { frame =>
+
+      val now = Time.now
+      // This future chain will march to completion even if the frameF future is discarded.  This
+      // guarantees that the entire stream will be consumed.
+      val frameF = stream.read().flatMap { frame =>
         if (frame.isEnd) {
           Future.value(Some(frame))
         } else {
           frame.release()
-          consumeAllButLast(stream)
+          consumeAllButLast(stream, deadline)
         }
       }
+
+      deadline match {
+        case Some(deadline) if deadline > now =>
+          // This is a sort of "safe" timeout.  If the classification deadline passes, we want to
+          // return a classificationTimeoutException but we _don't_ want to cancel the read Future.
+          // Cancelling the read or discarding the result of the read could result in reading a frame
+          // but never releasing it.  If we initiate a read, we must always capture the resulting
+          // frame and ensure it is released.
+          Future.selectIndex(IndexedSeq(Future.sleep(deadline - now), frameF)).flatMap {
```

Comment:
Yeah, I was able to figure out what was going on here, I just meant that the control flow with `selectIndex` is a little hard to follow at a glance?